File: packages/sdk-ts/test/spec-coverage.test.ts
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { OPERATIONS, SCOPES, VERSION, WixzelPhone } from '../src/index.js';
import { USER_AGENT } from '../src/version.js';

/**
 * The SDK is hand-written; this is what keeps it honest against the API.
 * Every operation in docs/openapi.json must map to exactly one method, every
 * method must name a real operation, and each must call what it claims.
 */
const here = dirname(fileURLToPath(import.meta.url));
const spec = JSON.parse(readFileSync(resolve(here, '../../../docs/openapi.json'), 'utf8')) as {
    paths: Record<string, Record<string, unknown>>;
    components: { schemas: { CreateApiKey: { properties: { scopes: { items: { enum: string[] } } } } } };
};
const pkg = JSON.parse(readFileSync(resolve(here, '../package.json'), 'utf8')) as { version: string; name: string };

const HTTP_METHODS = new Set(['get', 'post', 'put', 'patch', 'delete']);
const RESOURCES = [
    'engines', 'agents', 'calls', 'leads', 'campaigns', 'knowledgeBases',
    'phoneNumbers', 'sipTrunks', 'appointments', 'usage', 'billing', 'apiKeys',
] as const;

const client = new WixzelPhone({ apiKey: 'wv_test_x', fetch: (() => Promise.reject(new Error('no'))) as typeof fetch });

describe('spec coverage', () => {
    test('every OpenAPI operation has exactly one method, and every method names a real operation', () => {
        const operations = new Set<string>();
        for (const [path, methods] of Object.entries(spec.paths)) {
            for (const method of Object.keys(methods)) {
                if (HTTP_METHODS.has(method)) operations.add(`${method.toUpperCase()} ${path}`);
            }
        }
        const keys = Object.keys(OPERATIONS);
        const missing = [...operations].filter((op) => !keys.includes(op));
        const phantom = keys.filter((k) => !operations.has(k));
        const values = Object.values(OPERATIONS);
        const duplicates = values.filter((v, i) => values.indexOf(v) !== i);

        assert.deepEqual(missing, [], `operations in docs/openapi.json with no SDK method: ${missing.join(', ')}`);
        assert.deepEqual(phantom, [], `SDK table names an endpoint the spec does not have: ${phantom.join(', ')}`);
        assert.deepEqual(duplicates, [], `two operations map to the same method: ${duplicates.join(', ')}`);
        assert.equal(keys.length, operations.size);
    });

    test('each named method exists on the client and calls the verb and path it claims', () => {
        for (const [operation, name] of Object.entries(OPERATIONS)) {
            const [verb, path] = operation.split(' ') as [string, string];
            const [ns, method] = name.split('.') as [keyof WixzelPhone, string];
            const resource = client[ns] as unknown as Record<string, unknown>;
            const fn = resource?.[method];
            assert.equal(typeof fn, 'function', `${name} is not a method on the client`);
            const source = (fn as () => void).toString();
            // The transpiler may rewrite quotes, so accept either style. A
            // cursor list goes through http.list(), which is GET by definition.
            const sendsVerb = source.includes(`'${verb}'`) || source.includes(`"${verb}"`) || (verb === 'GET' && source.includes('.list('));
            assert.ok(sendsVerb, `${name} should send ${verb}`);
            const staticPrefix = path.split('{')[0]!.replace(/\/$/, '');
            assert.ok(source.includes(staticPrefix), `${name} should reference ${staticPrefix}`);
        }
    });

    test('no resource has a public method the table does not know about', () => {
        const known = new Set(Object.values(OPERATIONS) as string[]);
        for (const ns of RESOURCES) {
            const proto = Object.getPrototypeOf(client[ns]) as object;
            for (const method of Object.getOwnPropertyNames(proto)) {
                if (method === 'constructor') continue;
                assert.ok(known.has(`${ns}.${method}`), `${ns}.${method} is public but not in OPERATIONS`);
            }
        }
        assert.equal(RESOURCES.length, new Set(Object.values(OPERATIONS).map((v) => v.split('.')[0])).size);
    });

    test('the scope list matches the API enum exactly', () => {
        assert.deepEqual([...SCOPES], spec.components.schemas.CreateApiKey.properties.scopes.items.enum);
    });

    test('the version constant matches package.json', () => {
        assert.equal(VERSION, pkg.version);
        assert.equal(pkg.name, 'wixzel-phone');
        assert.ok(USER_AGENT.startsWith('wixzel-phone/'));
    });
});

File: packages/sdk-dart/test/spec_coverage_test.dart
@TestOn('vm')
library;

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';
import 'package:wixzel_phone/wixzel_phone.dart';

/// The SDK is hand-written; this is what keeps it honest against the API.
void main() {
  final specFile = File('../../docs/openapi.json');
  final client = WixzelPhone(apiKey: 'wv_test_x');

  group('spec coverage', () {
    late Map<String, dynamic> spec;
    late Set<String> operations;

    setUpAll(() {
      if (!specFile.existsSync()) {
        return;
      }
      spec = jsonDecode(specFile.readAsStringSync()) as Map<String, dynamic>;
      const httpMethods = {'get', 'post', 'put', 'patch', 'delete'};
      operations = {
        for (final entry in (spec['paths'] as Map<String, dynamic>).entries)
          for (final method in (entry.value as Map<String, dynamic>).keys)
            if (httpMethods.contains(method))
              '${method.toUpperCase()} ${entry.key}',
      };
    });

    test('every operation has exactly one method, and every method is real',
        () {
      if (!specFile.existsSync()) {
        return markTestSkipped('docs/openapi.json not reachable');
      }
      final missing =
          operations.where((op) => !kOperations.containsKey(op)).toList();
      final phantom =
          kOperations.keys.where((op) => !operations.contains(op)).toList();
      final values = kOperations.values.toList();
      final duplicates =
          values.where((v) => values.where((o) => o == v).length > 1).toSet();

      expect(missing, isEmpty,
          reason: 'operations with no SDK method: $missing');
      expect(phantom, isEmpty,
          reason: 'SDK names an endpoint the spec lacks: $phantom');
      expect(duplicates, isEmpty,
          reason: 'two operations share a method: $duplicates');
      expect(kOperations.length, operations.length);
    });

    test('every table value resolves to a real method', () {
      final methods = operationMethods(client);
      for (final name in kOperations.values) {
        expect(methods[name], isNotNull,
            reason: '$name is not wired in operationMethods');
      }
      for (final alias in kOperationAliases) {
        expect(methods[alias], isNotNull,
            reason: '$alias is not wired in operationMethods');
      }
      // Nothing in the mapping that the table and the aliases do not name.
      final known = {...kOperations.values, ...kOperationAliases};
      expect(methods.keys.where((k) => !known.contains(k)), isEmpty);
    });

    test('each resource file calls the verb and path its table entry claims',
        () {
      final dir = Directory('lib/src/resources');
      if (!dir.existsSync()) {
        return markTestSkipped('resources not reachable');
      }
      final sources = {
        for (final file in dir.listSync().whereType<File>())
          file.uri.pathSegments.last.replaceAll('.dart', ''):
              file.readAsStringSync(),
      };
      // resource name in the table -> file name, e.g. knowledgeBases -> knowledge_bases
      String fileFor(String resource) => resource.replaceAllMapped(
          RegExp('[A-Z]'), (m) => '_${m[0]!.toLowerCase()}');

      for (final entry in kOperations.entries) {
        final parts = entry.key.split(' ');
        final verb = parts[0];
        final path = parts[1];
        final resource = entry.value.split('.').first;
        final source = sources[fileFor(resource)];
        expect(source, isNotNull, reason: 'no source file for $resource');
        final staticPrefix =
            path.split('{').first.replaceAll(RegExp(r'/$'), '');
        expect(source, contains(staticPrefix),
            reason: '${entry.value} should reference $staticPrefix');
        // A cursor list goes through transport.list(), which is GET by definition.
        final sendsVerb = source!.contains("'$verb'") ||
            (verb == 'GET' && source.contains('.list('));
        expect(sendsVerb, isTrue, reason: '${entry.value} should send $verb');
      }
    });

    test('the scope enum matches the API exactly', () {
      if (!specFile.existsSync()) {
        return markTestSkipped('docs/openapi.json not reachable');
      }
      final schemas = (spec['components'] as Map<String, dynamic>)['schemas']
          as Map<String, dynamic>;
      final createApiKey = schemas['CreateApiKey'] as Map<String, dynamic>;
      final properties = createApiKey['properties'] as Map<String, dynamic>;
      final scopes = (properties['scopes'] as Map<String, dynamic>)['items']
          as Map<String, dynamic>;
      final expected = (scopes['enum'] as List<dynamic>).cast<String>();
      final actual = ApiScope.values
          .where((s) => s != ApiScope.unknown)
          .map((s) => s.value)
          .toList();
      expect(actual, expected);
    });

    test('the version constant matches pubspec.yaml', () {
      final pubspec = File('pubspec.yaml');
      if (!pubspec.existsSync()) {
        return markTestSkipped('pubspec.yaml not reachable');
      }
      final match = RegExp(r'^version:\s*(.+)$', multiLine: true)
          .firstMatch(pubspec.readAsStringSync());
      expect(match?.group(1)?.trim(), packageVersion);
    });
  });
}
